plugins/qcheck-stm/src/stm_of_ir.ml

```diff
@@ -676,29 +728,36 @@
   in
   let* rhs =
     let name = str_of_ident value.id in
-    let* pp_args =
-      concat_map
-        (fun (ty, id) ->
+    let rec aux ty args =
+      match (ty.ptyp_desc, args) with
+      | Ptyp_arrow (_, l, r), xs when Cfg.is_sut config l ->
+          let* fmt, pps = aux r xs in
+          ok ("%s" :: fmt, estring "sut" :: pps)
+      | Ptyp_arrow (_, _, r), (ty, id) :: xs ->
           let ty = subst_core_type value.inst ty in
-          let* pp = pp_of_ty ty in
-          ok [ pexp_apply pp [ (Nolabel, ebool true) ]; get_name id ])
-        value.args
-    in
-    let fmt =
-      String.concat " " ("%s" :: List.map (Fun.const "%a") value.args)
-      |> estring
+          let* pp = pp_of_ty ty and* fmt, pps = aux r xs in
+          ok
+            ( "%a" :: fmt,
+              pexp_apply pp [ (Nolabel, ebool true) ] :: get_name id :: pps )
+      | _, [] -> ok ([], [])
+      | _, _ ->
+          failwith
+            "shouldn't happen (list of arguments should be consistent with \
```

Comment:
I wonder if we shouldn’t create an error value `ShouldNotHappen` (in a separate PR if at all), as it appears in many already monadic contexts.